File: Implementations/Dining Savages/dining_savages.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <unistd.h>
#include <chrono>

int constexpr NUM_SAVAGES = 5;
int constexpr TOTAL_SERVINGS = 3000;
int constexpr TOTAL_REFILLS = 5;

struct Pot {
  std::mutex mutex;
  std::condition_variable pot_full;
  std::condition_variable pot_empty;
  int remaining_servings = TOTAL_SERVINGS;
};

struct Savage {
  int index;
};

int main() {
  auto begin = std::chrono::steady_clock::now();

  Pot pot;
  std::vector<std::thread> threads(NUM_SAVAGES + 1);

  threads[0] = std::thread ([&pot] () mutable {
    auto num_refills = 0;
    while (true) {
      std::unique_lock<std::mutex> lock(pot.mutex);
      pot.pot_empty.wait(lock);
      pot.pot_full.notify_all();

      ++num_refills;

      if (num_refills == TOTAL_REFILLS) {
        break;
      }
    }
  });


  for (auto i = 0; i < NUM_SAVAGES; i++) {
    Savage savage {i};

    threads[i + 1] = std::thread ([savage, &pot] () mutable {
      while (true) {
        std::unique_lock<std::mutex> lock(pot.mutex);
        if (pot.remaining_servings == 0) {
          // std::cout << "Telling cook to refill pot " << pot.remaining_servings << " from " << savage.index << std::endl;
          pot.pot_empty.notify_one();
          pot.pot_full.wait(lock);
          pot.remaining_servings = TOTAL_SERVINGS;
        }

        --pot.remaining_servings;
        // usleep(1e+3);
        // std::cout << "Remaining servings left in pot are " << pot.remaining_servings << " from " << savage.index << std::endl;
      }
    });
  }

  threads[0].join();

  auto end = std::chrono::steady_clock::now();
  std::cout << "Time difference = " << std::chrono::duration_cast<std::chrono::microseconds>(end - begin).count() << "µs" << std::endl;
}

File: Implementations/Dining Savages/dining_savages.go
package main

import (
  "fmt"
  "sync"
  "time"
)

const numSavages = 5
const totalServings = 3000
const totalRefills = 3

type Pot struct {
  mutex sync.Mutex
  remaining_servings int

  pot_empty chan bool
  pot_full chan bool
}

type Savage struct {
  index int
}

func main() {
  start := time.Now()
  var eating sync.WaitGroup

  pot_empty := make(chan bool)
  pot_full := make(chan bool)
  pot := Pot{remaining_servings: totalServings, pot_empty: pot_empty, pot_full: pot_full}

  eating.Add(1)
  go func(pot *Pot, eating *sync.WaitGroup) {
    num_refills := 0
    for {
      <- pot.pot_empty
      pot.pot_full <- true
      num_refills++

      if num_refills == totalRefills {
        eating.Done()
      }
    }
  }(&pot, &eating)

  for i := 0; i < numSavages; i++ {
    savage := Savage{i}

    go func(savage *Savage, pot *Pot, eating *sync.WaitGroup) {
      for {
        pot.mutex.Lock()
        if pot.remaining_servings == 0 {
          // fmt.Printf("Telling cook to refill pot %v from %v\n", pot.remaining_servings, savage.index)
          pot.pot_empty <- true
          <- pot.pot_full
          pot.remaining_servings = totalServings
        }
        pot.remaining_servings--
        // time.Sleep(5 * time.Millisecond)
        // fmt.Printf("Remaining servings left in pot are %v from %v savage\n", pot.remaining_servings, savage.index)
        pot.mutex.Unlock()
      }
    }(&savage, &pot, &eating)
  }

  eating.Wait()
  fmt.Printf("Time difference = %s\n", time.Since(start))
}
